mmgen/models/ddpm/denoising.py

```diff
@@ -0,0 +1,400 @@
+                image_size) == 2, 'The length of `image_size` should be 2.'
+            assert image_size[0] == image_size[
+                1], 'Width and height of the image should be same.'
+            image_size = image_size[0]
+        self.image_size = image_size
+
+        channels_cfg_ = deepcopy(self._default_channels_cfg)
+        if channels_cfg is not None:
+            channels_cfg_ = channels_cfg_.update(channels_cfg)
+        if isinstance(channels_cfg_, dict):
+            if image_size not in channels_cfg_:
+                raise KeyError(f'`image_size={image_size} is not found in '
+                               '`channel_cfg`, only support configs for '
+                               f'{[chn for chn in channels_cfg_.keys()]}')
+            self.channel_factor_list = channels_cfg_[image_size]
+        elif isinstance(channels_cfg_, list):
+            self.channel_factor_list = channels_cfg_
+        else:
+            raise ValueError('Only support list or dict for `channel_cfg`, '
+                             f'receive {type(channels_cfg_)}')
+
+        embedding_channels = base_channels * 4 \
+            if embedding_channels == -1 else embedding_channels
+        self.time_embedding = build_module(
+            dict(type='TimeEmbedding'), {
+                'in_channels': base_channels,
+                'embedding_channels': embedding_channels
+            })
+        if self.num_classes != 0:
+            self.label_embedding = nn.Embedding(self.num_classes,
+                                                embedding_channels)
+
+        self.resblock_cfg = deepcopy(resblock_cfg)
+        self.resblock_cfg.setdefault('dropout', dropout)
+        self.resblock_cfg.setdefault('norm_cfg', norm_cfg)
+        self.resblock_cfg.setdefault('act_cfg', act_cfg)
+        self.resblock_cfg.setdefault('embedding_channels', embedding_channels)
+        self.resblock_cfg.setdefault('use_scale_shift_norm',
+                                     use_scale_shift_norm)
+        self.resblock_cfg.setdefault('shortcut_kernel_size',
+                                     shortcut_kernel_size)
+
+        # factor to apply attention
+        attention_scale = [image_size // int(res) for res in attention_res]
+        self.attention_cfg = deepcopy(attention_cfg)
+        self.attention_cfg.setdefault('num_heads', num_heads)
+        self.attention_cfg.setdefault('norm_cfg', norm_cfg)
+
+        self.downsample_cfg = deepcopy(downsample_cfg)
+        self.downsample_cfg.setdefault('with_conv', downsample_conv)
+        self.upsample_cfg = deepcopy(upsample_cfg)
+        self.upsample_cfg.setdefault('with_conv', upsample_conv)
+
+        # init the factor
+        scale = 1
+        self.in_blocks = nn.ModuleList([
+            EmbedSequential(
+                nn.Conv2d(in_channels, base_channels, 3, 1, padding=1))
+        ])
+        self.in_channels_list = [base_channels]
+        for level, factor in enumerate(self.channel_factor_list):
```

Comment:
Is this `for-loop` for the encoder part in U-Net? If it is, leave a comment: `#  construct the encoder part`.

Reply:
Comment has added.